File: src/cxx-dev/LiENa/lienaRealTime/lienaDecodingTask.cpp
#include "lienaDecodingTask.h"


/**
 *
 * @brief igtDecodingTask::igtDecodingTask
 * @param dataware
 * @param inputQueueManager
 * @param datagramAnalyser
 */
lienaDecodingTask::lienaDecodingTask(lienaInputQueue *inputQueue,
                                     bool motivate,
                                     uint32_t localDeviceid,
                                     uint32_t targetDeviceId){

    this->motivate = motivate;
    this->localDeviceid = localDeviceid;
    this->targetDeviceId = targetDeviceId;
    this->inputQueue = inputQueue;

    this->decoder = new lienaDatagramDecoder(this->motivate, this->localDeviceid, this->targetDeviceId);
    this->rtFlag = true;
    this->rtPeriod = 20;
    this->standby = true;

    this->connect(this->decoder, SIGNAL(handShakeMessageArrived(lienaHandShakeMessage *)),                     this, SLOT(notiftHandShakeMessage(lienaHandShakeMessage *)));
    this->connect(this->decoder, SIGNAL(handShakeCommitMessageArrived()),                                      this, SLOT(channelOpen()));
    this->connect(this->decoder, SIGNAL(disengagementMessageArrived(lienaDisengagementMessage *)),             this, SLOT(notiftDisengagementMessage(lienaDisengagementMessage *)));
    this->connect(this->decoder, SIGNAL(disengagementCommitMessageArrived(lienaDisengagementCommitMessage *)), this, SLOT(notiftDisengagementCommitMessage(lienaDisengagementCommitMessage *)));
    this->connect(this->decoder, SIGNAL(connectionConfirm()),                                                  this, SLOT(connectionEstablish()));
    this->connect(this->decoder, SIGNAL(channelOpenMsg(lienaChannelOpenedMessage *)),                          this, SLOT(notiftHeartBeatMessage(lienaChannelOpenedMessage *)));
    this->connect(this->decoder, SIGNAL(channelClosedMessageArrived(lienaChannelClosedMessage * )),            this, SLOT(notifyChannelClosedMessage()));
    this->connect(this->decoder, SIGNAL(heartBeatMessageArrived(lienaHeartBeatMessage*)),                      this, SLOT(notifyHeartBeatMessage(lienaHeartBeatMessage*)));
    this->connect(this->decoder, SIGNAL(reHandshakeMessageArrived(lienaReHandshakeMessage*)),                  this, SLOT(notifyReHandshakeMsg(lienaReHandshakeMessage*)));
    this->connect(this->decoder, SIGNAL(reHandshakeCommitMessageArrived(lienaReHandshakeCommitMessage*)),      this, SLOT(notifyReHandshakeCommit(lienaReHandshakeCommitMessage*)));
    this->connect(this->decoder, SIGNAL(channelReOpenedMessageArrived(lienaChannelReOpened*)),                 this, SLOT(notifyChannelReOpened(lienaChannelReOpened*)));
    this->connect(this->decoder, SIGNAL(passiveNTPClockSynchronizationMessageArrived(lienaNetworkQualityMessage*)), this, SLOT(notifyPassiveNetworkQualityMessage(lienaNetworkQualityMessage*)));
    this->connect(this->decoder, SIGNAL(motivateNTPClockSynchronizationMessageArrived(lienaNetworkQualityMessage*)),this,SLOT(notifyMotivateNetworkQualityMessage(lienaNetworkQualityMessage*)));
}

//! ---------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDecodingTask::notifyMotivateNetworkQualityMessage
//! \param msg
//!
void lienaDecodingTask::notifyMotivateNetworkQualityMessage(lienaNetworkQualityMessage *msg){
    if(DEBUG){
        qDebug()<<"lienaDecodingTask::notifyMotivateNetworkQualityMessage";
    }
    emit networkMotivateQualityMessageArrived(msg);
}

//! ---------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDecodingTask::notifyPassiveNetworkQualityMessage
//! \param msg
//!
void lienaDecodingTask::notifyPassiveNetworkQualityMessage(lienaNetworkQualityMessage *msg){
    if(DEBUG){
        qDebug()<<"lienaDecodingTask::notifyPassiveNetworkQualityMessage";
    }
    emit networkPassiveQualityMessageArrived(msg);
}

//! ---------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDecodingTask::notifyReHandshakeMsg
//! \param msg
//!
void lienaDecodingTask::notifyReHandshakeMsg(lienaReHandshakeMessage* msg){
    if(DEBUG){
        qDebug()<<"lienaDecodingTask::notifyReHandshakeMsg";
    }
    emit rehandshakeMsgArrived(msg);
}

//! ---------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDecodingTask::notifyReHandshakeCommit
//! \param msg
//!
void lienaDecodingTask::notifyReHandshakeCommit(lienaReHandshakeCommitMessage* msg){
    emit rehandshakeCommitMsgArrived(msg);
}

//! ---------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDecodingTask::notifyChannelReOpened
//! \param msg
//!
void lienaDecodingTask::notifyChannelReOpened(lienaChannelReOpened* msg){
    emit reOpenedMsgArrived(msg);
}

//! ---------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDecodingTask::notifyHeartBeatMessage
//! \param msg
//!
void lienaDecodingTask::notifyHeartBeatMessage(lienaHeartBeatMessage* msg){
    emit heartBeatMessageArrived(msg);
}

//! ---------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDecodingTask::channelClosed
//!
void lienaDecodingTask::notifyChannelClosedMessage(){
    qDebug()<<"channelClosed";
    emit channelClosedMessageArrived();
}

//! ---------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDecodingTask::channelOpen
//!
void lienaDecodingTask::channelOpen(){
    qDebug()<<"lienaDecodingTask::channelOpen()";
    emit handShakeCommitMessageArrived();
}

//! ---------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDecodingTask::notiftHeartBeatMessage
//! \param channelOpenedMsg
//!
void lienaDecodingTask::notiftHeartBeatMessage(lienaChannelOpenedMessage* channelOpenedMsg){
    emit channelOpenMsg(channelOpenedMsg);
}

//! ---------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDecodingTask::~lienaDecodingTask
//!
lienaDecodingTask::~lienaDecodingTask(){
    delete this->decoder;
}

//! ---------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDecodingTask::enable
//!
void lienaDecodingTask::enable(){
    this->standby = false;

}

//! ---------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDecodingTask::freeze
//!
void lienaDecodingTask::freeze(){
    this->standby = true;

}

//! ---------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDecodingTask::launch
//!
void lienaDecodingTask::launch(){
    this->standby = false;
    this->start();
}

//! ---------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief igtDecodingTask::run
//!
void lienaDecodingTask::run(){
    while(this->rtFlag){

        if(standby){
            this->sleep(1);
            continue;
        }

        //! to check if the input queue is empty
        if(this->inputQueue->getLength() > 0){

            //! fetch
            lienaDatagram *datagram = inputQueue->getFrontDatagram();

            //! analysis
            this->decoder->analyse(datagram);

            //! delete
            this->inputQueue->deleteFrontDatagram();

        }

        //! latency
        msleep(this->rtPeriod);
    }
}

//! -----------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief igtDecodingTask::receptConnection
//! \param remoteIP
//!
void lienaDecodingTask::notiftHandShakeMessage(lienaHandShakeMessage *handShakeMessage){
    emit handShakeMessageArrived(handShakeMessage);
}

//! -----------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDecodingTask::notiftDisengagementMessage
//! \param disengagementMessage
//!
void lienaDecodingTask::notiftDisengagementMessage(lienaDisengagementMessage *disengagementMessage){
    emit disengagementMessageArrived(disengagementMessage);
}

void lienaDecodingTask::notiftDisengagementCommitMessage(lienaDisengagementCommitMessage *disengagementCommitMessage){
    qDebug()<<"lienaDecodingTask | notiftDisengagementCommitMessage emit disengagementCommitMessageArrived";
    emit disengagementCommitMessageArrived(disengagementCommitMessage);
}

//! ----------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDecodingTask::setPeriod
//! \param rtPeriod
//!
void lienaDecodingTask::setPeriod(uint32_t rtPeriod){
    this->rtPeriod = rtPeriod;
}

//! ----------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDecodingTask::setInputCache
//! \param inputMessageCache
//!
void lienaDecodingTask::setInputMessageQueue(lienaMessageQueue* inputMessageQueue){
    //this->decoder->setInputCache(inputMessageCache);
}

//! ----------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief igtDecodingTask::connectionEstablish
//!
void lienaDecodingTask::connectionEstablish(){
    emit connexionConfirm();
}

//! ----------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief igtDecodingTask::stop
//!
void lienaDecodingTask::stop(){
    rtFlag = false;
}

File: src/pyDev/LiENa/LiENaRealTimeTask/LienaDecodingTask.py
# -*- coding: utf-8 -*-
import threading
import time
from multiprocessing import Lock
from PyQt5.QtCore import QObject, pyqtSignal
from LiENa.LiENaStructure.LiENaDatagram.LienaDecoder import LienaDecoder
from LiENa.LiENaStructure.LiENaMessage.LienaNetworkQualityMessage import LienaNetworkQualityMessage
from LiENa.LiENaStructure.LiENaMessage.LienaChannelReOpenedMessage import LienaChannelReOpenedMessage
from LiENa.LiENaStructure.LiENaMessage.LienaChannelClosedMessage import LienaChannelClosedMessage
from LiENa.LiENaStructure.LiENaMessage.LienaDisengagementMessage import LienaDisengagementMessage
from LiENa.LiENaStructure.LiENaMessage.LienaHandShakeCommitMessage import LienaHandShakeCommitMessage
from LiENa.LiENaStructure.LiENaMessage.LienaHeartbeatMessage import LienaHeartbeatMessage
from LiENa.LiENaStructure.LiENaMessage.LienaHandShakeMessage import LienaHandShakeMessage
from LiENa.LiENaStructure.LiENaMessage.LienaChannelOpenedMessage import LienaChannelOpenedMessage
from LiENa.LiENaStructure.LiENaMessage.LienaReHandshakeCommitMessage import LienaReHandshakeCommitMessage
from LiENa.LiENaStructure.LiENaMessage.lienaReHandshakeMessage import LienaReHandshakeMessage
from LiENa.LiENaStructure.LiENaMessage.LienaDisengagementCommitMessage import LienaDisengagementCommitMessage
from LiENa.LiENaStructure.LiENaMessage.LienaCustomizedMessage import LienaCustomizedMessage
from LiENa.LiENaBasic.lienaDefinition import *
from RCPContext.LienaControlInstruction import LienaControlInstruction


class LienaDecodingTask(QObject):

    connectionConfirm = pyqtSignal()
    handshakeMessageArrived = pyqtSignal(LienaHandShakeMessage)
    rehandshakeMessageArrived = pyqtSignal(LienaReHandshakeMessage)
    rehandshakeCommitMessageArrived = pyqtSignal(LienaReHandshakeCommitMessage)
    channelReOpenedMessageArrived = pyqtSignal(LienaChannelReOpenedMessage)
    heartBeatMessageArrived = pyqtSignal(LienaHeartbeatMessage)
    disengagementMessageArrived = pyqtSignal(LienaDisengagementMessage)
    disengagementCommitMessageArrived = pyqtSignal(LienaDisengagementCommitMessage)
    channelClosedMessageArrived = pyqtSignal(LienaChannelClosedMessage)
    channelOpenedMessageArrived = pyqtSignal(LienaChannelOpenedMessage)
    handshakeCommitMessageArrived = pyqtSignal(LienaHandShakeCommitMessage)
    motivateNTPClockSynchronisationMessageArrived = pyqtSignal(LienaNetworkQualityMessage)
    passiveNTPClockSynchronisationMessageArrived = pyqtSignal(LienaNetworkQualityMessage)
    controlMessageArrived = pyqtSignal(LienaCustomizedMessage)

    def __init__(self, input_queue, motivate, local_device_id, target_device_id):
        super(LienaDecodingTask, self).__init__()
        self.inputQueue = input_queue
        self.motivate = motivate
        self.localDeviceId = local_device_id
        self.targetDeviceId = target_device_id

        self.inputMsgQue = None

        self.decoder = LienaDecoder(motivate, local_device_id, target_device_id)
        self.decoder.channelReOpenedMessageArrived[LienaChannelReOpenedMessage].connect(self.notify_channel_reopened_message_arrived)
        self.decoder.handshakeMessageArrived[LienaHandShakeMessage].connect(self.notify_handshake_message)
        self.decoder.handshakeCommitMessageArrived[LienaHandShakeCommitMessage].connect(self.notify_handshake_commit_message)
        self.decoder.channelOpenedMessageArrived[LienaChannelOpenedMessage].connect(self.notify_channel_opened_message)
        self.decoder.disengagementCommitMessageArrived[LienaDisengagementCommitMessage].connect(self.notify_disengagement_commit_message)
        self.decoder.disengagementMessageArrived[LienaDisengagementMessage].connect(self.notify_disengagement_message)
        self.decoder.channelClosedMessageArrived[LienaChannelClosedMessage].connect(self.notify_channel_closed_message)
        self.decoder.heartBeatMessageArrived[LienaHeartbeatMessage].connect(self.notify_heartbeat_message)
        self.decoder.rehandshakeMessageArrived[LienaReHandshakeMessage].connect(self.notify_rehandshake_message)
        self.decoder.rehandshakeCommitMessageArrived[LienaReHandshakeCommitMessage].connect(self.notify_rehandshake_commit_message)
        self.decoder.passiveNTPClockSynchronisationMessageArrived[LienaNetworkQualityMessage].connect(self.notify_passive_ntp_clock_synchronisation_message)
        self.decoder.motivateNTPClockSynchronisationMessageArrived[LienaNetworkQualityMessage].connect(self.notify_motivate_ntp_clock_synchronisation_message)
        self.decoder.customizedMessageArrived[LienaCustomizedMessage].connect(self.notify_control_instruction)

        self.rtFlag = True
        self.rtPeriod = 0.02
        self.stand_by = False
        self.receptionTask = threading.Thread(None, self.decode)

    def set_input_msg_queue(self, inputMsgQue):
        # print("123")
        # self.decoder.set_input_cache(inputMsgQue)
        self.inputMsgQue = inputMsgQue

    def enable(self):
        self.stand_by = False

    def freeze(self):
        self.stand_by = True

    def notify_control_instruction(self, msg):
        self.inputMsgQue.append(msg)

    def notify_channel_reopened_message_arrived(self, msg):
        if DEBUG:
            print("LienaDecodingTask | notifyChannelReOpenedMessageArrived")
        self.channelReOpenedMessageArrived.emit(msg)

    def notify_motivate_ntp_clock_synchronisation_message(self, msg):
        if DEBUG:
            print("LienaDecodingTask | notify_motivate_ntp_clock_synchronisation_message")
        self.motivateNTPClockSynchronisationMessageArrived.emit(msg)

    def notify_passive_ntp_clock_synchronisation_message(self, msg):
        if DEBUG:
            print("LienaDecodingTask | notify_passive_ntp_clock_synchronisation_message")
        self.passiveNTPClockSynchronisationMessageArrived.emit(msg)

    def notify_rehandshake_commit_message(self, msg):
        if DEBUG:
            print("LienaDecodingTask | recovery_down | rehandshakeCommitMessageArrived")
        self.rehandshakeCommitMessageArrived.emit(msg)

    def notify_rehandshake_message(self, msg):
        self.rehandshakeMessageArrived.emit(msg)

    def notify_heartbeat_message(self, msg):
        self.heartBeatMessageArrived.emit(msg)

    def notify_handshake_commit_message(self, msg):
        self.handshakeCommitMessageArrived.emit(msg)

    def notify_channel_closed_message(self, msg):
        self.channelClosedMessageArrived.emit(msg)

    def notify_disengagement_message(self, msg):
        self.disengagementMessageArrived.emit(msg)

    def notify_disengagement_commit_message(self, msg):
        self.disengagementCommitMessageArrived.emit(msg)

    def notify_channel_opened_message(self, msg):
        self.channelOpenedMessageArrived.emit(msg)

    def notify_handshake_message(self, msg):
        if DEBUG:
            print("LienaDecodingTask | notify_handshake_message")

        self.handshakeMessageArrived.emit(msg)

    def launch(self):
        self.receptionTask.start()

    def decode(self):
        while self.rtFlag:

            if self.stand_by:
                time.sleep(1)
                continue
            if self.inputQueue.get_length() > 0:
                datagram = self.inputQueue.get_latest_array()
                self.decoder.analyse(datagram)
            time.sleep(self.rtPeriod)

    def set_period(self, rt_period):
        self.rtPeriod = rt_period

    def set_input_cache(self, input_message_cache):
        self.decoder.set_input_cache(input_message_cache)

    def connection_established(self):
        self.connectionConfirm.emit()

    def terminate(self):
        self.rtFlag = False

    def get_handshake_commit_message(self):
        self.decoder.get_handshake_commit_message()
